06x-new-kafka-roller.md

```diff
@@ -0,0 +1,350 @@
+
+Depending on the observed states, the roller will perform specific actions. Those actions should cause a subsequent observation to cause a state transition. This iterative process continues until each node's state aligns with the desired state.
+
+It will also introduce an algorithm that can restart brokers in parallel when safety conditions are not violated. These conditions guarantee Kafka producer availability and cause minimal impact on controllers and overall stability of clusters.
+
+### Node State
+When a new reconciliation starts up, a context object is created for each node to store the state and other useful information used by the roller. It will have the following fields:
+
+ - <i>nodeRef</i>: NodeRef object that contains Node ID.
+ - <i>currentNodeRole</i>: Currently assigned process roles for this node (e.g. controller, broker).
+ - <i>lastKnownState</i>: It contains the last known state of the node based on information collected from the abstracted sources (Kubernetes API, KafkaAgent and Kafka Admin API). The table below describes the possible states.
+ - <i>restartReason</i>: It is updated based on the current predicate logic from the `Reconciler`. For example, an update in the Kafka CR is detected.
+ - <i>numRestartAttempts</i>: The value is incremented each time the node has been attempted to restart.
+ - <i>numReconfigAttempts</i>: The value is incremented each time the node has been attempted to reconfigure.
+ - <i>numRetries</i>: The value is incremented each time the node cannot be restarted/reconfigured due to not meeting safety conditions (more on this later).
+ - <i>lastTransitionTime</i>: System.nanoTime of last observed state transition.
+
+ The following table illustrates the proposed node states and the possible transitions:
+ | State            | Description      | Possible transitions |
+ | :--------------- | :--------------- | :----------- |
+ | UNKNOWN               | The initial state when creating `Context` for a node. We expect to transition from this state fairly quickly after creating the context for nodes.  | `NOT_RUNNING` `NOT_READY` `RECOVERING` `SERVING` |
+ | NOT_RUNNING           | Node is not running (Kafka process is not running). This is determined via Kubernetes API, more details for it below. | `NOT_READY`, `NOT_RUNNING`. | `RESTARTED` `SERVING` | 
+ | NOT_READY             | Node is running but not ready to serve requests which is determined by Kubernetes readiness probe (broker state < 2 OR == 127 OR controller is not listening on port). | `RESTARTED` `SERVING` |
```

Comment:
Where do the states such as init container, PodInitialization etc. fit in between these two states? In that state, the Pod is scheduled and some parts of it are running. But not the Kafka process. So it seems to fall between these two states right now.

Reply:
I believe this would fall to `NOT_READY` category. NOT_RUNNING has a specific criteria, which is very similar to the criteria of stuck pods in the current roller.

Reply:
I think that makes sense. But please add it to the state explanation as right now it suggests it would not belong there.

Reply:
There is already `NOT_RUNNING` section right below the table, explaining this.